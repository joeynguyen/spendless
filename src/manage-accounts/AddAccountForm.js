import React, { PropTypes } from 'react';
import { Button, Collapse, Input } from 'react-bootstrap';

const AddAccountForm = ({ doSubmit, fields, submitting }) => {
  // handleSubmit and fields are from redux-form
  const { accountName, accountType, accountCompany } = fields;
  let accountTypeField = '';
  if (accountType.value === 'bank') {
    accountTypeField = (
      <Input
        type="text"
        label="Name of Institution"
        placeholder="Enter the name of the financial institution"
        bsStyle={accountCompany.touched && accountCompany.invalid ? 'error' : null}
        help={accountCompany.touched ? accountCompany.error : ''}
        {...accountCompany}
      />
    );
  } else if (accountType.value === 'creditcard') {
    accountTypeField = (
      <Input
        type="select"
        label="Credit Card Company"
        placeholder="Credit Card Company"
        bsStyle={accountCompany.touched && accountCompany.invalid ? 'error' : null}
        help={accountCompany.touched ? accountCompany.error : ''}
        {...accountCompany}
      >
        <option value="">select...</option>
        <option value="Visa">Visa</option>
        <option value="MasterCard">Mastercard</option>
        <option value="American Express">American Express</option>
        <option value="Discover">Discover</option>
        <option value="Diners Club">Diners Club</option>
        <option value="JCB">JCB</option>
        <option value="Other">Other</option>
      </Input>
    );
  }

  return (
    <form onSubmit={doSubmit}>
      <Input
        type="text"
        label="Name"
        placeholder="Enter a name for the account"
        bsStyle={accountName.touched && accountName.invalid ? 'error' : null}
        help={accountName.touched ? accountName.error : ''}
        {...accountName}
      />

      <Input
        type="select"
        label="Type"
        placeholder="Type"
        bsStyle={accountType.touched && accountType.invalid ? 'error' : null}
        help={accountType.touched ? accountType.error : ''}
        {...accountType}
      >
        <option value="">select...</option>
        <option value="bank">Bank</option>
        <option value="creditcard">Credit Card</option>
      </Input>

      <Collapse in={accountType.value !== ''}>
        <div>{accountTypeField}</div>
      </Collapse>

      <Button
        bsStyle="primary"
        type="submit"
        disabled={submitting}>
        {submitting ? 'Saving...' : 'Save'}
      </Button>
    </form>
  );
};

AddAccountForm.propTypes = {
  doSubmit: PropTypes.func.isRequired,
  fields: PropTypes.object.isRequired,
  submitting: PropTypes.bool.isRequired,
};

export default AddAccountForm;
