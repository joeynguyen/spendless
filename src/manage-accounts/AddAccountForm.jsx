import PropTypes from 'prop-types';
import React from 'react';
import { Animate, Button } from 'grommet';

import FieldGroup from '../custom-components/FieldGroup.jsx';
import { accountCompanyOptions, accountTypeOptions } from '../constants.js';

const AddAccountForm = ({ localHandleSubmit, fields, submitting }) => {
  // fields and submitting are from redux-form
  const { accountName, accountType, accountCompany } = fields;
  let accountTypeField = '';
  if (accountType.value === 'bank') {
    accountTypeField = (
      <FieldGroup
        type="text"
        placeholder="Enter the name of the financial institution"
        label="Name of Institution"
        name={accountCompany.name}
        error={accountCompany.error}
        invalid={accountCompany.invalid}
        touched={accountCompany.touched}
        onBlur={accountCompany.onBlur}
        onChange={accountCompany.onChange}
        onFocus={accountCompany.onFocus}
        value={accountCompany.value}
      />
    );
  } else if (accountType.value === 'creditcard') {
    accountTypeField = (
      <FieldGroup
        componentClass="select"
        label="Credit Card Company"
        options={accountCompanyOptions}
        name={accountCompany.name}
        error={accountCompany.error}
        invalid={accountCompany.invalid}
        touched={accountCompany.touched}
        onBlur={accountCompany.onBlur}
        onChange={accountCompany.onChange}
        onFocus={accountCompany.onFocus}
        value={accountCompany.value}
      />
    );
  }

  return (
    <form id="add-account-form" onSubmit={localHandleSubmit}>
      <FieldGroup
        type="text"
        placeholder="Enter a name for the account"
        label="Name"
        name={accountName.name}
        error={accountName.error}
        invalid={accountName.invalid}
        touched={accountName.touched}
        onBlur={accountName.onBlur}
        onChange={accountName.onChange}
        onFocus={accountName.onFocus}
        value={accountName.value}
      />

      <FieldGroup
        componentClass="select"
        label="Type"
        options={accountTypeOptions}
        name={accountType.name}
        error={accountType.error}
        invalid={accountType.invalid}
        touched={accountType.touched}
        onBlur={accountType.onBlur}
        onChange={accountType.onChange}
        onFocus={accountType.onFocus}
        value={accountType.value}
      />

      <Animate
        visible={accountType.value !== ''}
        enter={{animation: 'slide-down', duration: 300, delay: 0}}
      >
        {accountTypeField}
      </Animate>

      <Button
        name="add-account"
        type="submit"
        accent
        label={submitting ? 'Saving...' : 'Save'}
        disabled={submitting}
      />
    </form>
  );
};

AddAccountForm.propTypes = {
  localHandleSubmit: PropTypes.func.isRequired,
  fields: PropTypes.object.isRequired,
  submitting: PropTypes.bool.isRequired,
};

export default AddAccountForm;
