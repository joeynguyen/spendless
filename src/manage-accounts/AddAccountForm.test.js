import expect from 'expect';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import AddAccountForm from './AddAccountForm.jsx';

const props = {
  doSubmit: () => null,
};

describe('AddAccountForm', () => {
  const wrapper = shallow(<AddAccountForm {...props} />);

  it('renders accountName field with empty value', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('renders 3 user input fields', () => {
    xit('renders accountName field with empty value', () => {
      expect(wrapper.find('input[name="accountName"]').length).toBe(1);
      // expect(wrapper.find('input[name="accountName"]').props().value).toBe('');
    });

    xit('renders accountType field with empty value', () => {
      expect(wrapper.find('select[name="accountType"]').length).toBe(1);
      // expect(wrapper.find('select[name="accountType"]').props().value).toBe('');
    });

    xit("doesn't render accountCompany field initially", () => {
      expect(wrapper.find('input[name="accountCompany"]').length).toBe(0);
    });
  });

  describe('on accountType change', () => {
    xit('accountCompany field should have an input element if accountType is bank', () => {
      // Input uses this.refs.input.getValue() which makes it hard to test
      // retest this after changing to FormControl which uses e.target.value
      // wrapper.find('select[name="accountType"]').simulate('change', {target: {value: 'bank'}});
      // expect(wrapper.find('select[name="accountType"]').prop('value')).toBe('bank');
      // expect(wrapper.find('input[name="accountCompany"]').length).toBe(1);
      // expect(wrapper.find('input[name="accountCompany"]').some('input')).toBe(true);
    });
  });
});
