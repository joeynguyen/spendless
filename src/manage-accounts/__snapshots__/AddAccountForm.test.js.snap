// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAccountForm renders accountName field with empty value 1`] = `
<AddAccountForm
  doSubmit={[Function]}
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
/>
`;
